cmd/src/repos_add_kvp.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/src-cli/internal/api"
+)
+
+func init() {
+	usage := `
+Examples:
+
+  Add a key-value pair to a repository:
+
+    	$ src repos add-kvp -repo=repoID -key=mykey -value=myvalue
+
+  Omitting -value will create a tag (a key with a null value).
+`
+
+	flagSet := flag.NewFlagSet("add-kvp", flag.ExitOnError)
+	usageFunc := func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'src repos %s':\n", flagSet.Name())
+		flagSet.PrintDefaults()
+		fmt.Println(usage)
+	}
+	var (
+		repoFlag  = flagSet.String("repo", "", `The ID of the repo to add the key-value pair to (required)`)
+		keyFlag   = flagSet.String("key", "", `The name of the key to add (required)`)
+		valueFlag = flagSet.String("value", "", `The value associated with the key. Defaults to null.`)
+		apiFlags  = api.NewFlags(flagSet)
+	)
+
+	handler := func(args []string) error {
+		if err := flagSet.Parse(args); err != nil {
+			return err
+		}
+		if *repoFlag == "" {
+			return errors.New("error: repo is required")
+		}
+
+		keyFlag = nil
+		valueFlag = nil
+		flagSet.Visit(func(f *flag.Flag) {
+			if f.Name == "key" {
+				key := f.Value.String()
+				keyFlag = &key
+			}
+
+			if f.Name == "value" {
+				value := f.Value.String()
+				valueFlag = &value
+			}
+		})
```

Comment:
Can't you just use `keyFlag` and `valueFlag` above?

Reply:
I thought I could, but it turns out that when you define flags using `flagset.String()` ([these lines](https://github.com/sourcegraph/src-cli/pull/896/files#diff-ae7d3d85e9e477ad6a3b6fe781d1637f29d2aae7bee90f3671c8be8679cb196eR29-R34)), it requires you to provide a default value for the flag, and that can't be `nil` for strings. So I'm setting the empty string `""` as the default value for the flags, which seems to be the convention elsewhere in this repo. But this means that if you do `src repos add-kvp -repo=myrepo` (omitting `-key` and `-value`) it'll create a KVP of the form `"":""` which is probably not what the user wants. This also means that without this call to `flagset.Visit()` there's no way for a user to create a key with a null value.

So using `flagset.Visit()` this way helps avoid surprises for the user by erroring out if they don't explicitly set `-key`. And if they want to create a tag they can just omit `-value` and it'll be created with a null value. If they _do_ want to create a key that's an empty string then they can do that by explicitly setting that flag's value to an empty string (like `-key=`)

Reply:
Aah, gotcha. That makes sense. Thanks for explaning!